src/ws/session-utils.c

```diff
@@ -726,6 +726,53 @@ utmp_log (int login,
     }
 }
 
+void
+btmp_log (const char *username,
+          const char *rhost)
+{
+  struct timeval tv;
+
+  /* the `tv` in the utmp struct is not actually a `struct timeval`, so
+   * we need to read into a temporary variable and then copy the fields.
+   */
+  gettimeofday (&tv, NULL);
+
+  struct utmp entry = {
+    .ut_line = "web console",
+    .ut_pid = getpid (),
+    .ut_tv.tv_sec = tv.tv_sec,
+    .ut_tv.tv_usec = tv.tv_usec,
+    .ut_type = LOGIN_PROCESS,
+  };
+
+  strncpy (entry.ut_host, rhost, sizeof entry.ut_host);
+  strncpy (entry.ut_user, username, sizeof entry.ut_user);
+
+  int fd = open (_PATH_BTMP, O_WRONLY | O_APPEND);
+  if (fd == -1)
+    {
+      warn ("open(%s) failed", _PATH_BTMP);
+      goto out;
+    }
+
+  ssize_t r = write (fd, &entry, sizeof entry);
+  if (r < 0)
+    {
+      warn ("write() %s failed", _PATH_BTMP);
+      goto out;
+    }
+  else if (r != sizeof entry)
+    {
+      warnx ("incomplete write() %s: %zu of %zu bytes",
+             _PATH_BTMP, r, sizeof entry);
```

Comment:
A partial write may leave the btmp file in a corrupted state. Could we try and truncate() to the original size if this happens?

Reply:
I'm actually not sure if this is true.  I've been reading some stuff about `O_APPEND` and `PIPE_BUF`, for example:

https://stackoverflow.com/questions/1154446/is-file-append-atomic-in-unix

This seems to suggest that writing less than `PIPE_BUF` to an `O_APPEND` file will either work or it won't.  In any case, if it does a partial write, we are in some very serious trouble, because the next writer might have already written the next record to it by the time we're in a state to do anything intelligent about it (and our "intelligent" efforts would probably just make things worse).

Also: nobody else seems to use locking with this file, so using it ourselves won't help things very much...

One other writer, openssh, does attempt to recover from partial writes by writing again.  I'm not sure this is the correct approach (for the above reasons), but I guess we could copy their behaviour...

Reply:
Update: I think the person writing that post might be wrong because I can't find any additional evidence for the claim they make about `PIPE_BUF` applying to files other than pipes.

It's a tricky problem, and I have no idea what we should properly do in this case... particularly in light of the fact that we may be interacting with other programs that are trying to take their own "corrective" actions at the same time as ours.

Reply:
Argh weird Unix semantics that a file can be opened multiple times in parallel for writing. I suppose if that happens, and if partial writes are a thing, then without locking there's nothing that we can do indeed. So let's ignore that then indeed.